### DISCUSSION TITLE
Bun with socket.io

### DISCUSSION THREAD

**Discussion by N0N1m3** [oven-sh/bun#2111] [Q&A] [answered]
Starting with version [4.4.0](https://socket.io/blog/socket-io-4-4-0/), a Socket.IO server can now bind to a [µWebSockets.js](https://github.com/uNetworking/uWebSockets.js) server.

```ts
const { App } = require("uWebSockets.js");
const { Server } = require("socket.io");

const app = new App();
const io = new Server();

io.attachApp(app);

io.on("connection", (socket) => {
  // ...
});

app.listen(3000, (token) => {
  if (!token) {
    console.warn("port already in use");
  }
});
```

As written in the Readme:  **The HTTP server and server-side _websockets_ are based on [uWebSockets](https://github.com/uNetworking/uWebSockets).**

So, is there any way to port this to the capabilities of bun?

P.S. **[Socket.io article](https://socket.io/docs/v4/server-installation/#usage-with-uwebsockets)**

**Comment by amytimed** [comment]
This is the main thing stopping me from using Bun at the moment, would love to see it.

**Reply by phederal** [reply]
If it's useful to anyone, I've created a socket.io implementation for Bun with full support for socket.io-client.
https://github.com/phederal/socket.io-bun

**Comment by jhwheeler** [comment]
Does socket.io work with Bun if you use the default `ws` instead of a µWebSockets.js server?

**Reply by Jarred-Sumner** [reply, marked as answer]
socket.io does work with the default `ws` and internally `ws` in Bun actually uses uWebSockets so there's no need to use the uWebSockets backend (even if the uWebSockets.js backend did work in Bun, which it does not)

**Reply by itsfuad** [reply]
How to work with socket.io in Bun?

I need to run socket.io + Hono api server in the same port. In nodejs we can use binding or just pass the http.creatServer() instance to the socket.io's server constructor. But looks like there are issues. If not, please show example.

**Reply by hepiska** [reply]
i try with uWebSockets that already default by elisyajs but seems not working for pub sub usecase

**Reply by LeulAria** [reply]
@itsfuad  have you found any example ?

**Reply by rgillan** [reply]
Unless you need the fallback (non ws) support in socket.io it is well worthwhile refactoring your application to use the native Bun websockets directly (Bun.serve etc.).  The underlying uWebsockets implementation in our experience will significantly perform better than ws, websocket and socket.io in nodejs implementations (indeed it was the reason we first started looking at Bun).  The pub/sub process works as well out of the box if you follow the Bun docs.

**Reply by AndriyAntonenko** [reply]
Are there any working examples here? I am using Hono, bun and trying to add socket io instead of native sockets

**Reply by luisvinicius09** [reply]
I'm also looking for examples

**Reply by empz** [reply]
I'm not sure why is it suggested to use native Bun websockets instead of socket.io. These are not the same thing. Socket.io has some extra functionality than native websockets do not have. It's not a drop-in replacement. The performance benefits are not always worth to refactor the whole application that's already using socket.io.

Can someone explain why socket.io can't be used with Bun? Is it because a missing Node.js API compatibility? If so, which one? Any idea on when will it be supported?

**Reply by hubdidub** [reply]
socket.io can be used with Bun, though.
I don't know if you can use it with Bun.serve, but if you use it as if you had a Node.js application, there are not problems. At least I have not encountered any

**Reply by empz** [reply]
Yeah, I've just noticed that. I don't need to use it with Bun.serve, just
like a regular Node.js app. Thanks.

**Comment by markwitt1** [comment]
Any update here?

**Comment by markwitt1** [comment]
Related: https://github.com/socketio/socket.io/issues/4969

**Comment by darrachequesne** [comment]
For future readers, we have implemented a low-level engine dedicated to Bun.

With Bun's HTTP server:

```js
import { Server as Engine } from "@socket.io/bun-engine";
import { Server } from "socket.io";

const io = new Server();

const engine = new Engine({
  path: "/socket.io/",
});

io.bind(engine);

io.on("connection", (socket) => {
  // ...
});

export default {
  port: 3000,
  idleTimeout: 30, // must be greater than the "pingInterval" option of the engine, which defaults to 25 seconds

  ...engine.handler(),
};
```

With Hono:

```js
import { Server as Engine } from "@socket.io/bun-engine";
import { Server } from "socket.io";
import { Hono } from "hono";

const io = new Server();

const engine = new Engine();

io.bind(engine);

io.on("connection", (socket) => {
  // ...
});

const app = new Hono();

const { websocket } = engine.handler();

export default {
  port: 3000,
  idleTimeout: 30, // must be greater than the "pingInterval" option of the engine, which defaults to 25 seconds

  fetch(req, server) {
    const url = new URL(req.url);

    if (url.pathname === "/socket.io/") {
      return engine.handleRequest(req, server);
    } else {
      return app.fetch(req, server);
    }
  },

  websocket
}
```

Source here: https://github.com/socketio/bun-engine

**Reply by KurtGokhan** [reply]
Is there an example how to do it in Elysia?

**Reply by darrachequesne** [reply]
Unfortunately, I was not able to make it work with Elysia, there seems to be no way to hook into the WebSocket upgrade...

**Reply by KurtGokhan** [reply]
Made a [small change](https://github.com/elysiajs/elysia/pull/1358) to Elysia for this.

For future readers, here is how to use it in the latest version of Elysia.

```ts
import { Elysia } from 'elysia';
import { Server as Engine } from '@socket.io/bun-engine';
import { Server } from 'socket.io';

const io = new Server();
const engine = new Engine({ path: '/socket.io/' });
io.bind(engine);

const app = new Elysia()
  .all('/socket.io/', ({ request, server }) => engine.handleRequest(request, server!))
  .listen({
    idleTimeout: 30,
    websocket: engine.handler().websocket as unknown as Bun.WebSocketHandler,
  });
```

**Reply by darrachequesne** [reply]
@KurtGokhan awesome! Added in the readme: https://github.com/socketio/bun-engine?tab=readme-ov-file#with-elysia
